### ISSUE TITLE
Questions about the codes in makeAby3Protocol

### ISSUE THREAD

**Issue by DuanYuFi** [secretflow/spu#58]
Hi, I'm recently learning how to create a new mpc protocol in spu. I read the `makeAby3Protocol` function in spu/mpc/aby3/protocol.cc: 29 and have some questions:
1. What is `aby3::registerTypes()` doing? Is it something about register the secret sharing type in arithmetic and binary?
2. What is `regABKernels()` doing? I saw and guess from the codes inside that it registers operations between plain and secret, but what's the difference with spu/mpc/aby3/protocol.cc:51 and below? (`obj->regKernel<aby3::P2A>()`, `obj->regKernel<aby3::AddAP>()`, etc.) Is this the relation like api and kernel? `regABKernels()` registers the api and for example `obj->regKernel<aby3::AddAP>()` is the underlying implementation.

Thanks for reply!

**Comment by 6fj** [comment]
Hi @DuanYuFi ,

1. basically correct. Please check spu/mpc/aby3/type.h for details.
2. You could see in both **spu/mpc/common/abprotocol.h** and **spu/mpc/aby3/protocol.h** we are registering ops. **spu/mpc/common/abprotocol.h** is shared between all protocols which own arithmetic and binary ops and conversions. And in the end, for a new protocol, you need to register all ops defined in **spu/mpc/api.h**.

**Comment by rivertalk** [comment]
Hi @DuanYuFi, thanks for the question.

1. `aby3::registerTypes()` tries to register `aby3` protocol related types to the global type registry, so we could use reflection on these types, i.e. create type instance from string representation: 
   https://github.com/secretflow/spu/blob/main/spu/core/type.cc#L101

2. `regABKernels()` just register a group of A/B related kernels to the object. i.e.
Inside `regABKernels()`, it registers a kernel named `mulss` via `obj->regKernel<ABProtMulSS>()`, which handles multiplication between secrets. In side `mulss`, it further dispatch to `mulaa`, which does multiplication between arithmetic shares.  
So in short, ABProtocol handles relationship between `Secret` and `Public`, and it dispatches `Secret` to `Arithmetic Share` or `Boolean Share` according to value's runtime information, which is registered in the concrete protocol (like ABY3).

Thanks,
Jin

**Comment by DuanYuFi** [comment]
> Hi @DuanYuFi ,
> 
> 1. basically correct. Please check spu/mpc/aby3/type.h for details.
> 2. You could see in both **spu/mpc/common/abprotocol.h** and **spu/mpc/aby3/protocol.h** we are registering ops. **spu/mpc/common/abprotocol.h** is shared between all protocols which own arithmetic and binary ops and conversions. And in the end, for a new protocol, you need to register all ops defined in **spu/mpc/api.h**.

oh thanks, I'll check type.h and api.h for details.

**Comment by DuanYuFi** [comment]
> Hi @DuanYuFi, thanks for the question.
> 
> 1. `aby3::registerTypes()` tries to register `aby3` protocol related types to the global type registry, so we could use reflection on these types, i.e. create type instance from string representation:
>    https://github.com/secretflow/spu/blob/main/spu/core/type.cc#L101
> 2. `regABKernels()` just register a group of A/B related kernels to the object. i.e.
>    Inside `regABKernels()`, it registers a kernel named `mulss` via `obj->regKernel<ABProtMulSS>()`, which handles multiplication between secrets. In side `mulss`, it further dispatch to `mulaa`, which does multiplication between arithmetic shares.
>    So in short, ABProtocol handles relationship between `Secret` and `Public`, and it dispatches `Secret` to `Arithmetic Share` or `Boolean Share` according to value's runtime information, which is registered in the concrete protocol (like ABY3).
> 
> Thanks, Jin

Thanks for reply, so here is what I understand: in the vision of `regABKernels()`, data sources are plain and secret, and it will determine which kind of secret it is (arithmetic share, binary share) and call the functions registered in the concrete protocol. Is that right?

**Comment by anakinxc** [comment]
> regABKernels

Yes, and also handles AB conversion when required. Like msb_s, if input is b share, it will convert to a share and forward to msb_a.

**Comment by DuanYuFi** [comment]
Ok, thanks everybody, it helps me a lot. :)
